### PR TITLE
Consistency fixes

### PR DESCRIPTION
**Pull request by sithmel**
This PR contains a few changes aimed to improve consistency in the library:

## new variadic curry function
Most of the functions are now using this wrapper. It helped to remove a lot of boilerplate:
* takes care of currying, and support variable arguments
* no need to wrap every iterable/asyncIterable in every function
* takes care of argument position. **Optional arguments go first**
* undefined is never a valid optional argument

## Breaking changes
~~undefined is not considered a "valid" iterable anymore~~
* takeSorted/asyncTakeSorted optional "comparator" argument is now the first argument

### CODE REVIEW

**Comment by sithmel** [comment]
related tickets: #188 #185

**Comment by KSXGitHub** [comment]
> undefined is not considered a "valid" iterable anymore

You meant "valid iterable element*"?

**Comment by conartist6** [comment]
I don't think it's an acceptable trade off for our methods to fail on undefined. That's one of the absolute biggest benefits of iter-tools: it should protect you from the whole class of undefined/null errors.

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Added
  - pipe
 
+### Changes
+ - takeSorted/asyncTakeSorted moved optional comparator argument to the beginning
```
Comment (conartist6) [inline CHANGELOG.md]: I recommend "optional comparator is now the first argument". Also I think you should also call out all the other arguments that can no longer be undefined here. My goal for this file is that if we change a method its name is in this file, such that if I search changelog for a method name I will know which releases have note-worthy changes to it.

Comment (sithmel) [reply CHANGELOG.md]: done

**Pull request merged**